lib/src/source/hosted.dart

```diff
@@ -226,6 +242,82 @@ class BoundHostedSource extends CachedSource {
     return result;
   }
 
+  /// If a cached version listing response for [ref] exists on disk and is less
+  /// than [maxAge] old it is parsed and returned.
+  ///
+  /// Otherwise deletes a cached response if it exists and returns `null`
+  Map<PackageId, _VersionInfo> _cachedVersionListingResponse(
+      PackageRef ref, Duration maxAge) {
+    final cachePath = _versionListingCachePath(ref);
+    final stat = io.File(cachePath).statSync();
+    final now = DateTime.now();
+    if (stat.type == io.FileSystemEntityType.file) {
+      if (now.difference(stat.modified) < maxAge) {
+        try {
+          final cachedDoc = jsonDecode(readTextFile(cachePath));
+          final timestamp = cachedDoc['timestamp'];
+          if (timestamp is String) {
+            final cacheAge =
+                DateTime.now().difference(DateTime.parse(timestamp));
+            if (cacheAge > maxAge) {
+              // Too old according to internal timestamp - delete.
+              tryDeleteEntry(cachePath);
+            } else {
+              return _versionInfoFromPackageListing(
+                  cachedDoc['response'], ref, Uri.file(cachePath));
+            }
+          }
+        } on io.IOException {
+          // Could not read the file. Delete if it exists.
+          tryDeleteEntry(cachePath);
+        } on FormatException {
+          // Decoding error - bad file or bad timestamp. Delete the file.
+          tryDeleteEntry(cachePath);
+        }
+      } else {
+        // File too old
+        tryDeleteEntry(cachePath);
+      }
+    }
+    return null;
+  }
+
+  /// Saves the (decoded) response from package-listing of [ref].
+  void _cacheVersionListingResponse(Object body, PackageRef ref) {
+    final path = _versionListingCachePath(ref);
+    try {
+      ensureDir(p.dirname(path));
+      writeTextFile(
+          path,
+          jsonEncode({
+            'timestamp': DateTime.now().toIso8601String(),
+            'response': body
+          }));
+    } on io.IOException catch (e) {
+      // Not being able to write this cache is not fatal. Just move on...
+      log.fine('Failed writing cache file. $e');
+    }
+  }
+
+  @override
+  Future<PackageStatus> status(PackageId id, Duration maxAge) async {
+    final ref = id.toRef();
+    // Did we already get info for this package?
+    var versionListing = _scheduler.peek(ref);
+    // Do we have a cached version response on disk?
+    versionListing ??= _cachedVersionListingResponse(ref, maxAge);
+    // Otherwise retrieve the info from the host.
+    versionListing ??= await _scheduler.schedule(ref);
+    return versionListing[id].status;
```

Comment:
Do we need to handle the case where `id` isn't present in `versionListing`...
Maybe it's better to return an empty `PackageStatus()`, rather than crashing?

```suggestion
    assert(versionListing[id] != null);
    if (versionListing[id] == null) {
      return PackageStatus(); // empty status is better than crashing.
    }
    return versionListing[id].status;
```

Reply:
> But I think for simplicity just returning the empty status might be relatively harmless here, and surely better than crashing.

Yeah, that's why I suggested an assertion, it shouldn't happen  -- and if it does we must be in one of those corner cases where a package has been removed, but still exists in users cache. And even that corner case is unlikely to work -- in any case, assert + avoid crash in production is probably a good balance.

Reply:
Done